File: leetcode_problems/typescript/src/lc2744_find_max_string_pairs.ts
function reverse_word(word: string): string {
  return word.split("").reverse().join("");
}

function maximumNumberOfStringPairs(words: string[]): number {
  let res: number = 0;
  let map: Map<string, number> = new Map();
  for (let word of words) {
    let rev = reverse_word(word);
    res += map.get(word) || 0;
    map.set(rev, (map.get(rev) || 0) + 1);
  }
  return res;
}

console.log(maximumNumberOfStringPairs(["cd", "ac", "dc", "ca", "zz"]));

File: leetcode_problems/cpp/lc2744_find_max_string_pairs.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <iostream>
using std::string;
using std::unordered_map;
using std::vector;

class Solution
{
public:
    string reverse(string word)
    {
        return string(word.rbegin(), word.rend());
    }

    int maximumNumberOfStringPairs(vector<string> &words)
    {
        unordered_map<string, int> map;
        int res = 0;
        for (string word : words)
        {
            string rev = reverse(word);
            res += map[word];
            map[rev]++;
        }
        return res;
    }
};

int main()
{
    vector<string> words = {"cd", "ac", "dc", "ca", "zz"};
    Solution sol;
    std::cout << sol.maximumNumberOfStringPairs(words) << std::endl;
    return 0;
}